### PR TITLE
Manage players through the `Game` class

### PR DESCRIPTION
**Pull request by ystros**
Previously, a lot of the work of adding players / spectators, managing
disconnecting, reconnecting and leaving games was being managed manually
through the socket handlers. This code has now been moved into the
`Game` class itself.

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -447,30 +435,12 @@ io.on('connection', function(socket) {
...
-
-                if(game.started && !game.finishedAt) {
-                    game.finishedAt = new Date();
-                    game.saveGame();
-                }                
-            }
+        runAndCatchErrors(game, () => {
+            game.leave(socket.request.user.username);
+            // Force update of game state to make sure spectators get disconnected.
+            socket.emit('gamestate', game.getState(socket.request.user.username));
```
Comment (ystros) [inline server/index.js]: Most of this refactor was just pushing existing code around, but this is an actual change. This used to only happen if a spectator left, but I didn't want to send in the socket / io object into game to only do it for spectators. I think it's fine (brief testing didn't come up with anything) but please review carefully.

File: server/game/game.js [hunk 2]
```diff
@@ -57,6 +58,10 @@ class Game extends EventEmitter {
         return player.constructor === Spectator;
     }
 
+    hasActivePlayer(playerName) {
+        return !!this.players[playerName] && !this.players[playerName].left;
```
Comment (cryogen) [inline server/game/game.js]: Checking for truthiness is fine here, rather than forcing it to bool, but I'm happy to leave it as is too

File: server/game/game.js [hunk 3]
```diff
@@ -829,14 +824,76 @@ class Game extends EventEmitter {
...
+            return false;
+        }
+
+        this.players[user.username] = new Player(socketId, user, this.owner === user.username, this);
+
+        return true;
+    }
+
+    isEmpty() {
+        return _.all(this.players, player => player.disconnected || player.left);
```
Comment (cryogen) [inline server/game/game.js]: How do you feel about renaming players to playersAndSpecatators?  Would you rather do it in a separate commit?

File: server/game/game.js [hunk 4]
```diff
@@ -829,14 +824,76 @@ class Game extends EventEmitter {
...
+    }
+
+    leave(playerName) {
+        var player = this.players[playerName];
+
+        if(!player) {
+            return;
+        }
+
+        this.addMessage('{0} has left the game', player);
```
Comment (cryogen) [inline server/game/game.js]: One idea i had recently, that your implementation now makes easier, is to notify specifically if it's a spectator that's left/disconnected - dunno about anyone else but i find myself having to check if it was my opponent who left or not :)

File: server/index.js (hunk 1 above)
Comment (cryogen) [inline server/index.js]: I actually think this is unnecessary as the game state is sent to all players and spectators below, if the game is still around by then

Comment (ystros) [inline server/index.js]: Just tested, this line is explicitly needed for spectators. This is because we delete the spectator from the players array, so when it updates the game state for everyone in the list they won't receive it (since they're not in the list). Without some sort of explicit update, the "Leave" button for spectators doesn't revert the UI back to the game list.

Comment (cryogen) [inline server/index.js]: That sounds familiar!  Probably why I added it in the first place :)

File: server/game/game.js (hunk 2 above)
Comment (ystros) [inline server/game/game.js]: In general, I try to explicitly return a bool for methods that have `has` or `is` in the name.

File: server/game/game.js (hunk 3 above)
Comment (ystros) [inline server/game/game.js]: Will do as part of this PR. 👍

File: server/game/game.js (hunk 4 above)
Comment (ystros) [inline server/game/game.js]: Seems like a good candidate for a separate PR / commit after this refactor.
